baseServer.go

```diff
@@ -99,16 +100,19 @@ func (serv *BaseServer[T]) GetAgentMap() map[uuid.UUID]T {
 }
 
 func (serv *BaseServer[T]) agentStoppedTalking(id uuid.UUID) {
-	serv.agentMapRWMutex.Lock()
-	serv.agentStoppedTalkingMap[id] = struct{}{}
-
-	if len(serv.agentStoppedTalkingMap) == len(serv.agentMap) {
-		serv.doneChannelOnce.Do(func() {
-			serv.messagingFinished <- struct{}{}
-			close(serv.messagingFinished)
-		})
+	if serv.asyncMessagingFlag {
+		serv.agentMapRWMutex.Lock()
+		serv.agentStoppedTalkingMap[id] = struct{}{}
+
+		if len(serv.agentStoppedTalkingMap) == len(serv.agentMap) {
+			serv.doneChannelOnce.Do(func() {
+				serv.messagingFinished <- struct{}{}
+				serv.asyncMessagingFlag = false
+				close(serv.messagingFinished)
+			})
+		}
+		serv.agentMapRWMutex.Unlock()
```

Comment:
I don't like heavily nested code. It's quite hard to read. 

Try:

```
if !serv.shouldRunAsyncMessaging {
    return
}
```

and you can run the subsequent code as before

Reply:
Also, the flag should be set to false in the `endMessaging` method of the server once `<-done` is received. this seems like a better separation of responsibility